Juhitha taking over for Aureliano imgur autoposter

AurHubertMax opened the pull request:
# Description
(PRIORITY HIGH) Jae: Create appropriate auto-poster for Imgur

## Related PRS (if any):
This frontend PR is related to the backend [PR 1296](https://github.com/OneCommunityGlobal/HGNRest/pull/1296)

## Main changes explained:
- Created folder `src/components/Announcements/ImgurComponents` to contain all necessary components
- Created folder `src/components/Announcements/ImgurComponents/helpers` for helper functions to post/schedule
- Created folder `src/components/Announcements/ImgurComponents/ImgurPostEditor` for imgur poster frontend logic and styles
- Created folder `src/components/Announcements/ImgurComponents/ImgurScheduledPostEditor` for imgur scheduled post frontend logic and styles
- Created file `src/components/Announcements/ImgurComponents/ImageUploader` for the image upload component
- Created file `src/components/Announcements/ImgurComponents/LoginButton` for the login to Imgur button component

## How to test:
### Initial setup to create Imgur account and application:
- Create Imgur Account:
  1. Go to [this link](https://imgur.com/register) to register an Imgur account
  2. Create a test account with a real, valid email **(Required)**
  3. Log in to Imgur and go to settings [here](https://imgur.com/account/settings)
  4. Verify your email, if email have not been sent, click on the `Resend email verification` button **(Important)**


- Modify Backend `.env` file:
  1. Put these variables below in the backend `.env` file:
  ```
  REACT_APP_IMGUR_CLIENT_ID2=
  REACT_APP_IMGUR_REDIRECT_URI=http://localhost:4500/api/imgur/auth-callback
  REACT_APP_IMGUR_CLIENT_SECRET=
  ```
  2. Message me on slack (Aureliano Hubert Maximus) to ask for IMGUR_CLIENT_ID2 and REACT_APP_IMGUR_CLIENT_SECRET value, and paste the ids into the REACT_APP_IMGUR_CLIENT_ID2 and REACT_APP_IMGUR_CLIENT_SECRET.

- Use ngrok to Host frontend:
  **This is needed because Imgur does not allow users to log in from http:// request, only from https://. Ngrok allows us to host the frontend in https:// for free, and safely**
  1. create an account/sign in to [ngrok](https://ngrok.com/) on the top right of the page.
  2. go to either this link for [windows](https://dashboard.ngrok.com/get-started/setup/windows) users or this link for [macOs](https://dashboard.ngrok.com/get-started/setup/macos)
  3. Download ngrok and run it once download is complete
  4. put this into the opened command line: `ngrok http http://localhost:3000`
  5. IMPORTANT: use ngrok's provided url to test, dont use the localhost, as it will not work. If you come across the error as seen at the last screen of the video below, it is only because your frontend is not running yet. Once it is running, reload the page and it will work fine.
  6. Tutorial video: 

https://github.com/user-attachments/assets/9f89feba-03be-4fa8-9ea5-4c76f1114ed0



### Testing the autoposter:
1. check into current branch
8. do `npm install` and `npm run start:local` to run this PR locally
9. Clear site data/cache
10. log as owner user
11. go to dashboard→Other Links→Send Emails→Toggle Imgur icon next to 'Weekly Progress Editor'
12. Verify functionalities below:
  - Create and upload a post immediately
      - Fill in the title, description, topic, and tags section
      - Upload an image
      - Click `Post Now` button
      - Verify in your Imgur account that the post is successfully uploaded, along with the correct title, images, descriptions, and tags

  - Create and upload a scheduled post
      - Repeat steps 1 to 4 in *Create and upload a post immediately*
      - Click the calendar button and schedule a date to be posted
      - Click `Schedule` button to schedule the post
      - Verify that the post exists under *Scheduled Posts* section after reloading the page
      - Verify that the images to be posted exist in Imgur account, but is not yet uploaded to a gallery (so its visibility is hidden)
      - Verify that the post will be posted to a gallery in your Imgur account on the specified time.

## Screenshots or videos of changes:

https://github.com/user-attachments/assets/e70ef97c-1da7-45fa-9302-762b8c8d8561


## Note:
- Please do not post any pictures/text/any information if it contains the API keys you've just made.

AurHubertMax committed:
Add Imgur image upload functionality to Announcements component

AurHubertMax committed:
Refactor Imgur posting functionality in Announcements component and remove deprecated postToImgur action

AurHubertMax committed:
Add scheduled posts functionality and Imgur post handling in Announcements component

AurHubertMax committed:
Remove unused deleteScheduledPost function and enhance Imgur post styling in Announcements component

AurHubertMax committed:
Refactor Imgur post styling and structure in Announcements component to accurately show a preview of the post

AurHubertMax committed:
Add styling and delete, edit functionality for scheduled Imgur posts in Announcements component

AurHubertMax committed:
Enhance Imgur post input fields with tooltips and improved styling in Announcements component

AurHubertMax committed:
Add Imgur post logging and validation checks in Announcements component

AurHubertMax committed:
Implement input validation for Imgur post details and remove unused code in Announcements component

palankigreeshma1109 commented:
Hi @AurHubertMax , just to let you know this is my first time working with Imgur Api , Followed all the steps that you mentioned but I am encountering an error while trying to post. Attached few screenshots for your reference.
1.Created an an Imgur account

![Screenshot (31)](https://github.com/user-attachments/assets/95896650-4d1e-4a10-bead-3b2d5e2b32a3)

2.Created Imgur application.

![Screenshot (30)](https://github.com/user-attachments/assets/8649c671-8c04-4836-9553-c44d6f0c10ae)

3.added the mentioned 3 variables to notepad and updated them in the .env file
Client ID, Client secret and refresh_token

![Screenshot (28)](https://github.com/user-attachments/assets/46099f73-3a17-4039-a557-55c18e2fc610)


![Screenshot (29)](https://github.com/user-attachments/assets/7fc7a042-b571-46e6-988e-14aef19acd0f)

4.Now logged in as owner and tried to post an image. This is the error encountered.

![Screenshot (32)](https://github.com/user-attachments/assets/2fe985a5-49db-4d6d-9bea-50a19adf80ba)

![Screenshot (33)](https://github.com/user-attachments/assets/ebaf423d-e3be-4d28-96db-91dae2f8ddfc)

and some code that may raise error from my backend bash file

Received request to post to Imgur: {
  body: [Object: null prototype] {
    title: 'test1',
    topic: 'Testing the Imgur',
    tags: 'Imgur',
    description: 'Testing the Imgur Api 1 Post'
  },
  scheduleTime: undefined,
  files: [ { originalname: 'onecommunity1.jpg', description: 'T' } ]
}
Updated request body: [Object: null prototype] {
  title: 'test1',
  topic: 'Testing the Imgur',
  tags: 'Imgur',
  description: 'Testing the Imgur Api 1 Post'
}
Received request to post to Imgur: [Object: null prototype] {
  title: 'test1',
  topic: 'Testing the Imgur',
  tags: 'Imgur',
  description: 'Testing the Imgur Api 1 Post'
}
Received files: [
  {
    fieldname: 'image',
    originalname: 'onecommunity1.jpg',
    encoding: '7bit',
    mimetype: 'image/jpeg',
    buffer: <Buffer ff d8 ff e0 00 10 4a 46 49 46 00 01 01 00 00 01 00 01 00 00 ff e1 00 18 45 78 69 66 00 00 49 49 2a 00 08 00 00 00 00 00 00 00 00 00 00 00 ff e2 01 d8 ... 44055 more bytes>,
    size: 44105,
    description: 'T'
  }

Let me know if any further validation or changes are needed!

AurHubertMax commented:
> Hi @AurHubertMax , just to let you know this is my first time working with Imgur Api , Followed all the steps that you mentioned but I am encountering an error while trying to post. Attached few screenshots for your reference. 1.Created an an Imgur account

Hi @palankigreeshma1109, I had to delete one of your screenshots because it contained your API keys. It is not safe for you to upload public API keys anywhere. I will look into the problem. Thank you for testing it!

manushajyasta30 reviewed (commented):
I have followed all the steps mentioned in the description and added all the tokens required to connect to imgur api in the env variables of backend application. 
Here are the few points I noticed:

1. I have filled all the fields, uploaded a image and added description to the image. and able to successfully schedule/post the image. After the image is posted, you are clearing all the input fields but the image file seems to persist. 
<img width="955" alt="image" src="https://github.com/user-attachments/assets/c460213b-7f6f-4c89-b31c-35cebbe75438" />

2. Since the previous file name is not removed from the page, I directly filled the text fields and try to post without uploading the picture which shows below error.
<img width="953" alt="image" src="https://github.com/user-attachments/assets/2ac8d01a-bb78-4623-9db5-5c9b7d3be445" />

3. I tried uploading the same image that posted previously which i'm seeing on the page, It seems to not upload successfully and preview image is not showing in the preview section.
Only when I choose a new picture, it's allowing to upload successfully.

Please let me know if you need further clarification on the points listed above

AurHubertMax committed:
Add Imgur image upload functionality with improved styling and refactor input handling in Announcements component

AurHubertMax commented:
> I have followed all the steps mentioned in the description and added all the tokens required to connect to imgur api in the env variables of backend application. Here are the few points I noticed:
> 
> 1. I have filled all the fields, uploaded a image and added description to the image. and able to successfully schedule/post the image. After the image is posted, you are clearing all the input fields but the image file seems to persist.
> 
> 2. Since the previous file name is not removed from the page, I directly filled the text fields and try to post without uploading the picture which shows below error.
> 
> 3. I tried uploading the same image that posted previously which i'm seeing on the page, It seems to not upload successfully and preview image is not showing in the preview section.
>    Only when I choose a new picture, it's allowing to upload successfully.
> 
> Please let me know if you need further clarification on the points listed above

Hi @manushajyasta30, thank you for reviewing this PR and giving me feedback! I've applied the fixes and have pushed the changes to this branch. Please let me know if the problem still persist. Otherwise, please either put either "Request Changes" if there are any more issues or "Approve" on this PR. 

![image](https://github.com/user-attachments/assets/0a7edd9c-4c21-4656-ae81-427069acf263)

palankigreeshma1109 reviewed (dismissed):
I have reviewed this PR and followed all the steps mentioned in this PR
1.Created Imgur account
2.Created Imgur application
3.Added IMGUR_CLIENT_ID, IMGUR_CLIENT_SECRET and IMGUR_REFRESH_TOKEN to the .env file of backend.
4.Loged in as owner and navigated to the branch go to dashboard→Other Links→Send Emails→Toggle Imgur icon next to 'Weekly Progress Editor and found that the following functionalities are working as expected.

Review:
1.Created and upload a post successfully 

![Screenshot (68)](https://github.com/user-attachments/assets/e3f1e416-1dc3-437e-b6ea-6df20b87ee25)

Posted image in Imgur account
![Screenshot (69)](https://github.com/user-attachments/assets/92488421-ba30-460a-ba66-130920bba9dd)

2.Created and scheduled a post successfully.
![Screenshot (70)](https://github.com/user-attachments/assets/4de4fed9-ed8c-4057-864f-6a3368c60278)

![Screenshot (72)](https://github.com/user-attachments/assets/cfe8b4e0-5285-402d-8467-3014cfcf7e77)

and also able to delete a scheduled post successfully.

![Screenshot (71)](https://github.com/user-attachments/assets/edd12b8d-75e0-4991-afcd-e6c38bc01525)

3.i icon are also working as expected.
Attached the screenshots for your verification.

AurHubertMax committed:
Add Imgur integration for posting and scheduling images

- Implemented ImgurPostHelpers for handling image uploads, authentication checks, and disconnection from Imgur.
- Created ImgurScheduledPostHelpers for fetching and scheduling posts on Imgur.
- Updated Announcements component to include Imgur post editor and manage scheduled posts.
- Refactored URL utility to include new Imgur API endpoints for authentication, image uploads, and scheduled posts.
- Enhanced error handling and user feedback with toast notifications during Imgur operations.

AurHubertMax committed:
fixed URL endpoints to reflect what's implemented in the backend

AurHubertMax committed:
Add title, topic, and tags inputs to Imgur post editor; enhance scheduled post display with title, topic, and tags

AurHubertMax committed:
removed console.log and fixed linting errors

AurHubertMax committed:
adjust color and styling of imgur button

AurHubertMax committed:
Merge branch 'development' into Aureliano_Imgur_autoposter

AurHubertMax committed:
implemented a temporary fix to a test case in 'src\components\Projects\WBS\WBSDetail\components\__tests__\TagsSearch.test.js'

AurHubertMax committed:
more linting fixes

Venk-rgb reviewed (changes requested):
Hi @AurHubertMax . I went through the test steps. I created an Imgur account, downloaded ngrok and also logged in with an owner account. But the UI I see is totally different from the one you have shared in the video. Am I missing something?
<img width="1470" alt="PR #3341 - 5" src="https://github.com/user-attachments/assets/39b1c6b4-288b-42d3-ad2c-d56d99d91715" />

boppanapraveen reviewed (commented):
I followed the steps mentioned above and created an Imgur account, but I'm not able to establish a connection to post the image. I might be missing something could someone please let me know if there's a step I'm overlooking?

![#3341](https://github.com/user-attachments/assets/2daeb1d8-f7dd-4313-aa79-2ae3313b028b)
![#3341-1](https://github.com/user-attachments/assets/c20fe4cc-bde2-4e58-98f0-9597008a84f9)

naiduajaykumar reviewed (commented):
Can you please share
REACT_APP_IMGUR_CLIENT_ID2 and REACT_APP_IMGUR_CLIENT_SECRET values, I also messaged you in slack but didn't get any response from you.

ShravyaKudlu reviewed (commented):
It says imgur is over capacity when I try to connect
<img width="1341" height="330" alt="Screenshot from 2025-07-31 18-33-35" src="https://github.com/user-attachments/assets/634b9c23-7cf7-4756-a566-461b97d15a4f" />

rohanrastogi311 reviewed (changes requested):
Hi Aureliano, 

Launch Error due to the wrong node version. 14 instead of 20. 

<img width="1728" height="264" alt="Image" src="https://github.com/user-attachments/assets/e557fc97-ac1c-455c-a815-b0c819cfd626" />

sawantgayatri19 reviewed (changes requested):
Hello @AurHubertMax, thank you for your work on this pull request! 

To help get this PR ready for final review and approval, please address the following feedback and issues raised across the review thread:

1. Resolve Merge Conflicts: There are active merge conflicts in the following files that must be resolved:
   - `.eslintignore`
   - `src/components/Announcements/Announcements.css`
   - `src/components/Announcements/index.jsx`
   - `src/components/Projects/WBS/WBSDetail/components/__tests__/TagsSearch.test.js`
   - `src/routes.js`
   - `src/utils/URL.js`

2. Node Version Requirement:** Ensure your documentation/testing notes specify using Node v20 (as older versions like v14 cause launch failures).

3. UI State & Input Bugs:
   - Please look into the issue where image files persist or fail to preview properly if subsequent posts are attempted without resetting states completely.
   - Address the UI discrepancies reported where components or views differ from the original walkthrough video.

4. API Setup/Credentials:Clarify how secondary reviewers can access or safely test with the required Imgur credentials (`REACT_APP_IMGUR_CLIENT_ID2` / `REACT_APP_IMGUR_CLIENT_SECRET`) without blocking testing.

Once conflicts are resolved and these items are checked, please update the PR so we can proceed with a fresh review. Thank you!
